api/bucketsd/service.go

```diff
@@ -961,6 +1007,223 @@ func (s *Service) createLinks(
+		return nil, fmt.Errorf("verifying bucket update: %v", err)
+	}
+
+	fromNodePath, err := inflateFilePath(buck, fromPth)
+	if err != nil {
+		return nil, err
+	}
+	fromItem, err := s.pathToItem(ctx, buck, fromNodePath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	toNodePath, err := inflateFilePath(buck, toPth)
+	if err != nil {
+		return nil, err
+	}
+
+	toItem, err := s.pathToItem(ctx, buck, toNodePath, false)
+	if err == nil {
+		if fromItem.IsDir && !toItem.IsDir {
+			return nil, fmt.Errorf("Not a directory")
+		}
+		if toItem.IsDir {
+			// from => to becomes cleanDst:
+			// "c" => "b" becomes "b/c"
+			// "a.jpg" => "b" becomes "b/a.jpg"
+			toPth = gopath.Join(toPth, fromItem.Name)
+		}
+	}
+
+	// bucket node
+	buckNode, err := s.getNodeAtPath(ctx, pth, buck.GetLinkEncryptionKey())
+	if err != nil {
+		return nil, fmt.Errorf("getting node: %v", err)
+	}
+
+	var dirPath path.Resolved
+	// Private buckets need to manage keys+encrypted nodes
+	if buck.IsPrivate() {
+		dirPath, err = s.copyNode(ctx, buck, buckNode, fromPth, toPth)
+		if err != nil {
+			return nil, fmt.Errorf("copying node: %v", err)
+		}
+	} else {
+		buckPath := path.New(buck.Path)
+		_, dirPath, err = s.setPathFromExistingPath(ctx, buck, buckPath, fromPth, toPth, buckNode.Cid(), nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("copying path: %v", err)
+		}
+	}
+	buck.Path = dirPath.String()
+
+	// If "a/b" => "a/", cleanup only needed for priv
+	if strings.HasPrefix(fromPth, toPth) {
+		if buck.IsPrivate() {
+			dirPath, err := s.removePath(ctx, dbID, dbToken, buck, fromPth)
+			if err != nil {
+				return nil, fmt.Errorf("cleaning path: %v", err)
+			}
+			buck.Path = dirPath.String()
+		}
+
+		if err = s.saveAndPublish(ctx, dbID, dbToken, buck); err != nil {
+			return nil, err
+		}
+		return &pb.MovePathResponse{}, nil
+	}
+
+	if strings.HasPrefix(toPth, fromPth) {
+		// If "a/" => "a/b" cleanup each leaf in "a" that isn't "b" (skipping .textileseed)
+		ppth := path.Join(path.New(buck.Path), fromPth)
+		item, err := s.listPath(ctx, dbID, dbToken, buck, ppth)
+		if err != nil {
+			return nil, fmt.Errorf("listing path: %v", err)
+		}
+		reg := regexp.MustCompile("/ipfs/([^/]+)/")
+		for _, chld := range item.Item.Items {
+			sp := cleanPath(reg.ReplaceAllString(chld.Path, ""))
+			if strings.Compare(chld.Name, buckets.SeedName) == 0 || sp == toPth {
+				continue
+			}
+			spClean := cleanPath(sp)
+			dirPath, err := s.removePath(ctx, dbID, dbToken, buck, spClean)
+			if err != nil {
+				return nil, fmt.Errorf("cleaning path: %v", err)
+			}
+			buck.Path = dirPath.String()
+		}
+	} else {
+		// if a/ => b/ remove a
+		dirPath, err := s.removePath(ctx, dbID, dbToken, buck, fromPth)
+		if err != nil {
+			return nil, fmt.Errorf("removing path: %v", err)
+		}
+		buck.Path = dirPath.String()
+	}
+
+	buck.UpdatedAt = time.Now().UnixNano()
+	buck.SetMetadataAtPath(toPth, tdb.Metadata{
+		UpdatedAt: buck.UpdatedAt,
+	})
+	buck.UnsetMetadataWithPrefix(fromPth)
```

Comment:
This needs to be done before calling Verify, otherwise it won't have metadata change to verify.